lib/smart_answer/resolve_state.rb

```diff
@@ -0,0 +1,80 @@
+module SmartAnswer
+  class ResolveState
+    def initialize(flow)
+      @flow = flow
+    end
+
+    def from_response_store(response_store, requested_node = nil)
+      forwarding_responses = flow.response_store == :session ? {} : response_store.all
+      state = start_state(forwarding_responses)
+      loop do
+        node_name = flow.node(state.current_node).name.to_s
+        return state unless response_store.has?(node_name)
+
+        response = response_store.get(node_name)
+        return current_state(state, response) if node_name == requested_node
+
+        new_state = next_state(state, response)
+        return new_state if new_state.error
+        return new_state if flow.node(new_state.current_node).outcome?
+
+        state = new_state
+      end
+    end
+
+    def from_params(params)
+      responses = params[:responses].to_s.split("/")
+
+      state = responses.inject(start_state) do |memo, response|
+        return memo if memo.error
+
+        next_state(memo, response)
+      end
+
+      if params[:next]
+        next_state(state, params[:response])
+      elsif params[:previous_response]
+        current_state(state, params[:previous_response])
+      else
+        state
+      end
+    end
```

Comment:
I'd not want to move the ResolveState logic into the class since that's pretty heavy duty but it'd be quite possible to have it abstracted away.

So something like:

```ruby
class SmartAnswer::State
  def self.resolve_from_params(flow, params)
    ResolveState.new(flow).from_params(params)
  end

  def self.resolve_from_response_store(flow, response_store, requested_node: nil)
    ResolveState.new(flow).from_response_store(response_store, requested_node)
  end
end
```

Then change code controller code like: 

```ruby
  def show
    state = SmartAnswer::ResolveState.new(flow).from_response_store(response_store, node_name)
    @presenter = FlowPresenter.new(flow, state)
```

to:

```ruby
  def show
    state = SmartAnswer::State.resolve_from_response_store(flow, response_store, requested_node: node_name)
    @presenter = FlowPresenter.new(flow, state)
```

is that what you mean?

Reply:
Ah no, I was thinking more along the lines of:
```ruby
module SmartAnswer
  class State < OpenStruct
    def initialize(start_node, forwarding_responses: {})
       super(current_node: start_node,
                  accepted_responses: {},
                  forwarding_responses: forwarding_responses,
                  current_response: nil,
                  error: nil)
    end

    def self.new_from_params(flow, params)
      ...
    end

    def self.new_from_response_store(flow, response_store, requested_node = nil)
      ...
    end
    ...
end
```

It's not too much a sticking point for me and agree it's a lot of complicated logic to go in a object constructor. 

If we do go down the route of keeping it in a separate class - I can't see any benefit in having wrapper methods in the State class (your example), so would be fine calling the original method directly.